packages/@aws-cdk/aws-rds/lib/cluster-engine.ts

```diff
@@ -0,0 +1,318 @@
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as iam from '@aws-cdk/aws-iam';
+import * as secretsmanager from '@aws-cdk/aws-secretsmanager';
+import * as core from '@aws-cdk/core';
+import { ClusterParameterGroup, IParameterGroup, ParameterGroup } from './parameter-group';
+import { ParameterGroupFamilyMapping } from './private/parameter-group-family-mapping';
+import { compare } from './private/version';
+
+/**
+ * The extra options passed to the {@link IClusterEngine.bindToCluster} method.
+ */
+export interface ClusterEngineBindOptions {
+  /**
+   * The role used for S3 importing.
+   *
+   * @default - none
+   */
+  readonly s3ImportRole?: iam.IRole;
+
+  /**
+   * The role used for S3 exporting.
+   *
+   *  @default - none
+   */
+  readonly s3ExportRole?: iam.IRole;
+
+  /**
+   * The customer-provided ParameterGroup.
+   *
+   * @default - none
+   */
+  readonly parameterGroup?: IParameterGroup;
+}
+
+/**
+ * The type returned from the {@link IClusterEngine.bindToCluster} method.
+ */
+export interface ClusterEngineConfig {
+  /**
+   * The ParameterGroup to use for the cluster.
+   *
+   * @default - no ParameterGroup will be used
+   */
+  readonly parameterGroup?: IParameterGroup;
+
+  /**
+   * The port to use for this cluster,
+   * unless the customer specified the port directly.
+   *
+   * @default - use the default port for clusters (3306)
+   */
+  readonly port?: number;
+}
+
+/**
+ * The interface representing a database cluster (as opposed to instance) engine.
+ */
+export interface IClusterEngine {
+  /** The type of the engine, for example "aurora-mysql". */
+  readonly engineType: string;
+
+  /**
+   * The exact version of a given engine.
+   *
+   * @default - default version for the given engine type
+   */
+  readonly engineVersion?: string;
+
+  /**
+   * The family to use for ParameterGroups using this engine.
+   * This is usually equal to "<engineType><engineMajorVersion>",
+   * but can sometimes be a variation of that.
+   * You can pass this property when creating new ClusterParameterGroup.
+   */
+  readonly parameterGroupFamily: string;
+
+  /** The application used by this engine to perform rotation for a single-user scenario. */
+  readonly singleUserRotationApplication: secretsmanager.SecretRotationApplication;
+
+  /** The application used by this engine to perform rotation for a multi-user scenario. */
+  readonly multiUserRotationApplication: secretsmanager.SecretRotationApplication;
+
+  /** The default EC2 instance type to use for databases in this cluster. */
+  readonly defaultInstanceType: ec2.InstanceType;
+
+  /**
+   * Method called when the engine is used to create a new cluster.
+   */
+  bindToCluster(scope: core.Construct, options: ClusterEngineBindOptions): ClusterEngineConfig;
```

Comment:
Just `bind()` is sufficient and pass the `cluster` as a parameter of this method.

The API should be `bind(cluster: rds.ICluster, options: ClusterEngineBindOptions)`.

scope is an optional parameter of `ClusterEngineBindOptions` and defaults to `cluster`.

Reply:
> Just `bind()` is sufficient and pass the `cluster` as a parameter of this method.

1. There is a non-zero chance that there will be an engine in the future that can be used both for Instances and Clusters. If both interfaces have a `bind` method, that will make it impossible to implement both interfaces in one instance.
2. I want to avoid TypeScript's structural typing allowing you to pass an `IClusterEngine` where an `IInstanceEngine` is required, and vice versa.

> The API should be `bind(cluster: rds.ICluster, options: ClusterEngineBindOptions)`.
> 
> scope is an optional parameter of `ClusterEngineBindOptions` and defaults to `cluster`.

I had it take `IDatabaseCluster` (there's no `ICluster` interface in RDS) at the beginning. But the problem is that it doesn't really work. The problem is that `bindToCluster()` has to be called in the constructor of `DatabaseCluster`. Which means you'll actually pass a partially-constructed object to `bindToCluster()`, which will most likely result in errors when accessing any of its properties. In order to avoid those sort of errors, I think it's correct not to have `IDatabaseCluster` in the signature of `bindToCluster`.

Reply:
> There is a non-zero chance that there will be an engine in the future that can be used both for Instances and Clusters.

I disagree with this notion. This feels like premature optimization.
We should go with a model that - an engine is either a cluster engine or an instance engine. We shouldn't attempt to implement both at the same time in the same 'class'.
Some code duplication is ok, for the sake of code cleanliness. If it does end up that there is a significant piece of functionality common across them, we can move that logic into a separate package-private function and reuse. I'd rather lean towards a cleaner model.

> I want to avoid TypeScript's structural typing allowing you to pass an IClusterEngine where an IInstanceEngine is required, and vice versa

The `bind()` is going to be called once within the implementation of `ClusterEngine`. Anywhere else it will be invoked?
I'm not convinced this is a problem. How likely is this even? 
A more likely problem is that a user may (accidentally) pass a `ClusterEngine` as the `engine` property of a `DatabaseInstance`. I don't believe this code is preventing that, yes?

> The problem is that bindToCluster() has to be called in the constructor of DatabaseCluster. Which means you'll actually pass a partially-constructed object to bindToCluster(), which will most likely result in errors when accessing any of its properties.

Are you saying that we're doing this wrong everywhere? like [here](https://github.com/aws/aws-cdk/blob/99c3397d2018d8276907bde7fbe97c77d914cefa/packages/%40aws-cdk/aws-apigateway/lib/method.ts#L274) and [here](https://github.com/aws/aws-cdk/blob/99c3397d2018d8276907bde7fbe97c77d914cefa/packages/%40aws-cdk/aws-lambda/lib/function.ts#L502) and many more places I've seen this.

Reply:
> > There is a non-zero chance that there will be an engine in the future that can be used both for Instances and Clusters.
> 
> I disagree with this notion. This feels like premature optimization.
> We should go with a model that - an engine is either a cluster engine or an instance engine. We shouldn't attempt to implement both at the same time in the same 'class'.
> Some code duplication is ok, for the sake of code cleanliness. If it does end up that there is a significant piece of functionality common across them, we can move that logic into a separate package-private function and reuse. I'd rather lean towards a cleaner model.

I don't understand your point here, sorry. The issue is _not_ code duplication; we can hide it easily with this version, as you said. The issue is correct modeling.

If tomorrow, RDS adds support for the Oracle Standard Edition 3 engine, which is supported both for Clusters and Instances, I want to have a single class `OracleStandardEdition3Engine` which implements both `IClusterEngine` and `IInstanceEngine`, and that you can pass to both. What I _don't_ want to be forced to do is have two separate classes, `OracleSe3ClusterEngine` and `OracleSe3InstanceEngine` that have the same value for `engineType`, and have the same type for their engine version. That would be a modeling miss.

I don't see why the preference between `bind()` and `bindToCluster()` should make modeling this case correctly impossible. It's completely arbitrary what the name is; customers are not expected to call that method, ever.

Reply:
> > I want to avoid TypeScript's structural typing allowing you to pass an IClusterEngine where an IInstanceEngine is required, and vice versa
> 
> The `bind()` is going to be called once within the implementation of `ClusterEngine`. Anywhere else it will be invoked?
> I'm not convinced this is a problem. How likely is this even?
> A more likely problem is that a user may (accidentally) pass a `ClusterEngine` as the `engine` property of a `DatabaseInstance`. I don't believe this code is preventing that, yes?

Sorry, I have a little trouble following your point here, and I don't know what "this code" is referring to exactly, so let me reiterate what I wrote above.

If you have the following interfaces:

```ts
interface IClusterEngine {
  readonly engineType: string;
  readonly engineVersion: string;
  readonly parameterGroupFamily: string;
  bind(scope: Construct): string;
}

interface IInstanceEngine {
  readonly engineType: string;
  readonly engineVersion: string;
  readonly parameterGroupFamily: string;
  bind(scope: Construct): string;
}
```

Because of TypeScript's structural typing, you can pass an instance of `IClusterEngine` where the expected type is `IInstanceEngine`, and also pass an instance of `IInstanceEngine` where the expected type is `IClusterEngine`. So, this would be legal TypeScript:

```ts
new rds.DatabaseCluster(this, 'Cluster', {
  engine: rds.DatabaseInstanceEngine.MARIADB, // will fail at deploy time!
  // ...
});
```

If you change the names from `bind()` to `bindToCluster()` and `bindToInstance()`, this problem goes away, and the above code will compile (which is the goal here).

Reply:
> > The problem is that bindToCluster() has to be called in the constructor of DatabaseCluster. Which means you'll actually pass a partially-constructed object to bindToCluster(), which will most likely result in errors when accessing any of its properties.
> 
> Are you saying that we're doing this wrong everywhere? like [here](https://github.com/aws/aws-cdk/blob/99c3397d2018d8276907bde7fbe97c77d914cefa/packages/%40aws-cdk/aws-apigateway/lib/method.ts#L274) 

Yes, this is wrong. If the implementation of `Integration` calls, for example, the `methodId` property of the `Method` it has been passed, it will get `undefined`, even though its type is `string`, and will most likely crash with an exception at runtime. `bind` should have the signature `bind(scope: Construct, extraOptions: SomeOptions)` in this case, like it's done in this PR.

> and [here](https://github.com/aws/aws-cdk/blob/99c3397d2018d8276907bde7fbe97c77d914cefa/packages/%40aws-cdk/aws-lambda/lib/function.ts#L502) and many more places.

No, this is correct. Notice that `bind()` takes a `Construct` as the only argument. So this is correct.
> Even if so, has this problem showed up anywhere yet? Should we be solving for this now?

I don't understand your question. If you pass a half-constructed `IDatabaseCluster` or `IDatabaseInstance` to `bindToCluster()` / `bindToInstance()`, yes, this will be a problem.

Reply:
Fair enough. Point conceded.